framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java

```diff
@@ -894,56 +895,79 @@ public void report(final Result r, final Object src) {
-                        hasOption("warns")
-                                ? Diagnostic.Kind.MANDATORY_WARNING
-                                : Diagnostic.Kind.ERROR,
-                        src,
-                        msg.getMessageKey(),
-                        msg.getArgs());
-            } else if (r.isWarning()) {
-                this.message(
-                        Diagnostic.Kind.MANDATORY_WARNING, src, msg.getMessageKey(), msg.getArgs());
-            } else {
-                this.message(Diagnostic.Kind.NOTE, src, msg.getMessageKey(), msg.getArgs());
-            }
+        for (DiagMessage dmsg : r.getDiagMessages()) {
+            report(src, dmsg);
         }
     }
 
     /**
-     * Prints a message (error, warning, note, etc.) via JSR-269.
+     * Reports an error. By default, prints it to the screen via the compiler's internal messager.
      *
-     * @param kind the type of message to print
-     * @param source the object from which to obtain source position information; may be an Element,
-     *     a Tree, or null
-     * @param msgKey the message key to print
+     * @param source the source position information; may be an Element, a Tree, or null
+     * @param messageKey the message key
      * @param args arguments for interpolation in the string corresponding to the given message key
-     * @see Diagnostic
-     * @throws IllegalArgumentException if {@code source} is neither a {@link Tree} nor an {@link
-     *     Element}
-     */
-    private void message(
-            Diagnostic.Kind kind,
-            @Nullable Object source,
-            @CompilerMessageKey String msgKey,
-            Object... args) {
+     */
+    public void reportError(Object source, @CompilerMessageKey String messageKey, Object... args) {
+        report(source, ERROR, messageKey, args);
+    }
+
+    /**
+     * Reports a warning. By default, prints it to the screen via the compiler's internal messager.
+     *
+     * @param source the source position information; may be an Element, a Tree, or null
+     * @param messageKey the message key
+     * @param args arguments for interpolation in the string corresponding to the given message key
+     */
+    public void reportWarning(
+            Object source, @CompilerMessageKey String messageKey, Object... args) {
+        report(source, MANDATORY_WARNING, messageKey, args);
+    }
 
+    /**
+     * Reports a diagnostic message. By default, prints it to the screen via the compiler's internal
+     * messager.
+     *
+     * @param source the source position information; may be an Element, a Tree, or null
+     * @param d the diagnostic message
+     */
+    public void report(Object source, DiagMessage d) {
+        report(source, d.getKind(), d.getMessageKey(), d.getArgs());
+    }
+
+    /**
+     * Reports a diagnostic message. By default, it prints it to the screen via the compiler's
+     * internal messager; otherwise, the method returns with no side effects.
```

Comment:
I'm a bit confused about the `By default` and `otherwise` text. This is a private method, so subclasses can't adapt it. How/when do I reach the `otherwise` case?

Reply:
Via `printOrStoreMessage`.  I'll clarify the comment.